Nukambe/marci-forms
# Fix Non-Functional Delete Button for Questions

# Fix Non-Functional Delete Button for Questions

## Issue Addressed
This PR addresses the issue where the delete button for questions in the form builder was non-responsive. Users reported that clicking the delete button did not have any effect, leaving the question in place. Reference to the issue: #3 .

## Summary of Investigation and Fix
Upon investigating the issue, several steps were taken to diagnose and resolve the problem:
- Initial Attempt with useCallback: Initially suspected that the issue might be related to unnecessary re-renders or function recreations. Wrapping functions in useCallback was attempted but did not resolve the issue. Solution suggested via (https://stackoverflow.com/questions/77979278/why-is-this-react-state-variable-not-accepting-new-values)
- Component Conflict Discovery: Further investigation revealed that a nested component within the Question component was also utilizing context, which interfered with the expected behavior. Commenting out this nested component led to the normal functioning of the delete button.
- Resolution by Avoiding Context in Nested Component: To solve the issue, the use of context was removed for the specific scenario involving the problematic nested component. This change restored the expected functionality of the delete button.
- Deep Copying Arrays: Additionally, a significant improvement was made by changing the method of copying arrays from reference copying to deep copying using JSON.parse(JSON.stringify(array)). This approach ensures that arrays are copied by value, preventing unintended side effects caused by reference manipulation.

## How to Test
1. Navigate to the form builder section of the Marci Forms application.
2. Add new questions to a form or use an existing form with multiple questions.
3. Attempt to delete a question using the delete button.
4. Verify that the question is successfully removed from the form.
5. Observe that no adverse effects occur in the form builder's state or functionality.

## Impact of Changes
- User Experience: These fixes significantly enhance the form builder's usability, allowing users to efficiently manage their form questions.
- Application Stability: Addressing the underlying issue with context usage and array copying improves the overall stability and predictability of the application's state management.

## Additional Notes
- This fix highlights the importance of careful consideration when using context in nested components and managing state immutability through proper copying techniques.
- Further testing across different browsers and devices is recommended to ensure the robustness of the solution.
